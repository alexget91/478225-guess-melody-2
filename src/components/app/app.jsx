import React, {PureComponent} from "react";
import PropTypes from "prop-types";
import WelcomeScreen from "../welcome-screen/welcome-screen";
import GenreQuestionScreen from "../genre-question-screen/genre-question-screen";
import ArtistQuestionScreen from "../artist-question-screen/artist-question-screen";
import {artistQuestion, genreQuestion} from "../../common/global-prop-types";
import {connect} from "react-redux";
import {ActionCreator} from "../../reducer";
import GameScreen from "../game-screen/game-screen";

class App extends PureComponent {
  static getScreen(props) {
    const {step} = props;

    if (step === -1) {
      const {
        gameTime,
        maxMistakes,
        onWelcomeScreenClick
      } = props;

      return <WelcomeScreen
        time={gameTime}
        errorCount={maxMistakes}
        onStartButtonClick={onWelcomeScreenClick}
      />;
    }

    const {questions, mistakes, maxMistakes, onUserAnswer} = props;
    const currentQuestion = questions[step];

    switch (currentQuestion.type) {
      case `genre`: return <GameScreen mistakes={mistakes}>
        <GenreQuestionScreen
          screenIndex={step}
          question={currentQuestion}
          onAnswer={(userAnswer) => onUserAnswer(userAnswer, currentQuestion, mistakes, maxMistakes)}
        />
      </GameScreen>;

      case `artist`: return <GameScreen mistakes={mistakes}>
        <ArtistQuestionScreen
          screenIndex={step}
          question={currentQuestion}
          onAnswer={(userAnswer) => onUserAnswer(userAnswer, currentQuestion, mistakes, maxMistakes)}
        />
      </GameScreen>;
    }

    return null;
  }

  render() {
    return App.getScreen(this.props);
  }
}

App.propTypes = {
  mistakes: PropTypes.number.isRequired,
  maxMistakes: PropTypes.number.isRequired,
  gameTime: PropTypes.number.isRequired,
  questions: PropTypes.arrayOf(PropTypes.oneOfType([
    PropTypes.exact(genreQuestion),
    PropTypes.exact(artistQuestion),
  ])).isRequired,
  step: PropTypes.number.isRequired,
  onUserAnswer: PropTypes.func.isRequired,
  onWelcomeScreenClick: PropTypes.func.isRequired,
};

const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {
  step: state.step,
  mistakes: state.mistakes,
});

const mapDispatchToProps = (dispatch) => ({
  onWelcomeScreenClick: () => dispatch(ActionCreator.incrementStep()),

  onUserAnswer: (userAnswer, question, mistakes, maxMistakes) => {
    dispatch(ActionCreator.incrementStep());
    dispatch(ActionCreator.incrementMistakes(userAnswer, question, mistakes, maxMistakes));
  }
});

export {App};

export default connect(mapStateToProps, mapDispatchToProps)(App);
