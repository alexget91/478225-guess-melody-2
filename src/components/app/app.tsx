import * as React from "react";
import WelcomeScreen from "../welcome-screen/welcome-screen";
import GenreQuestionScreen from "../genre-question-screen/genre-question-screen";
import ArtistQuestionScreen from "../artist-question-screen/artist-question-screen";
import {connect} from "react-redux";
import {ActionCreator} from "../../reducer/game/reducer/reducer";
import GameScreen from "../game-screen/game-screen";
import FailScreen from "../fail-screen/fail-screen";
import withActivePlayer from "../../hocs/with-active-player/with-active-player";
import withUserAnswer from "../../hocs/with-user-answer/with-user-answer";
import {FailType, QuestionType, ScreenStep} from "../../common/constants";
import {getMistakes, getStep, getTime} from "../../reducer/game/selectors/selectors";
import {getQuestions} from "../../reducer/data/selectors/selectors";
import AuthorizationScreen from "../authorization-screen/authorization-screen";
import {getAuthorizationRequired, getUser} from "../../reducer/user/selectors/selectors";
import UserOperation from "../../reducer/user/operation/operation";
import {Switch, Route, Redirect} from "react-router-dom";
import Path from "../../common/path";
import WinScreen from "../win-screen/win-screen";
import {User, ArtistQuestion, GenreQuestion} from "../../common/types";

const GenreQuestionScreenWrapped = withUserAnswer(withActivePlayer(GenreQuestionScreen));
const ArtistQuestionScreenWrapped = withActivePlayer(ArtistQuestionScreen);

interface Props {
  mistakes?: number,
  maxMistakes?: number,
  gameTime: number,
  questions: Array<ArtistQuestion | GenreQuestion>,
  step: number,
  user?: User,
  isAuthorizationRequired?: boolean,
  onUserAnswer?: (answer: string | Array<boolean>, question: ArtistQuestion | GenreQuestion) => void,
  onWelcomeScreenClick?: () => void,
  onTimeChange?: () => void,
  onSignIn?: () => void,
  onReplayClick?: () => void,
}

const getScreen = (props: Props) => {
  const {questions, step, mistakes, maxMistakes, gameTime} = props;

  if (step >= questions.length) {
    return <Redirect to={Path.WIN}/>;
  }

  if (gameTime <= 0) {
    return <Redirect to={Path.LOSE_TIME}/>;
  }

  if (mistakes >= maxMistakes) {
    return <Redirect to={Path.LOSE_MISTAKES}/>;
  }

  if (step === ScreenStep.START) {
    const {onWelcomeScreenClick} = props;

    return <WelcomeScreen
      time={gameTime / 60}
      errorCount={maxMistakes}
      onStartButtonClick={onWelcomeScreenClick}
    />;
  }

  const {onUserAnswer, onTimeChange} = props;
  const currentQuestion = questions[step];

  switch (currentQuestion.type) {
    case QuestionType.GENRE: return <GameScreen type={currentQuestion.type} mistakes={mistakes} time={gameTime}
      onTimeChange={onTimeChange}>

      <GenreQuestionScreenWrapped
        screenIndex={step}
        question={currentQuestion}
        answersLength={currentQuestion.answers.length}
        onAnswer={(userAnswer) => onUserAnswer(userAnswer, currentQuestion)}
      />
    </GameScreen>;

    case QuestionType.ARTIST: return <GameScreen type={currentQuestion.type} mistakes={mistakes} time={gameTime}
      onTimeChange={onTimeChange}>

      <ArtistQuestionScreenWrapped
        screenIndex={step}
        question={currentQuestion}
        onAnswer={(userAnswer) => onUserAnswer(userAnswer, currentQuestion)}
      />
    </GameScreen>;
  }

  return null;
};

const App = (props: Props) => {
  const {user, onSignIn, onReplayClick} = props;

  return <Switch>
    <Route path={Path.INDEX} exact render={() => getScreen(props)}/>
    <Route path={Path.AUTH} exact render={() => <AuthorizationScreen
      onAuthFormSubmit={onSignIn}
    />}/>
    <Route path={Path.WIN} exact render={() => <WinScreen
      user={user}
      onAuthFormSubmit={onSignIn}
      onReplayClick={onReplayClick}
    />}/>
    <Route path={Path.LOSE_MISTAKES} exact render={() => {
      return <FailScreen failType={FailType.MISTAKES} onReplayClick={onReplayClick}/>;
    }}/>
    <Route path={Path.LOSE_TIME} exact render={() => {
      return <FailScreen failType={FailType.TIME} onReplayClick={onReplayClick}/>;
    }}/>
  </Switch>;
};


const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {
  questions: getQuestions(state),
  step: getStep(state),
  mistakes: getMistakes(state),
  gameTime: getTime(state),
  isAuthorizationRequired: getAuthorizationRequired(state),
  user: getUser(state),
});

const mapDispatchToProps = (dispatch) => ({
  onWelcomeScreenClick: () => dispatch(ActionCreator.incrementStep()),
  onUserAnswer: (userAnswer, question) => {
    dispatch(ActionCreator.incrementStep());
    dispatch(ActionCreator.incrementMistakes(userAnswer, question));
  },
  onTimeChange: () => dispatch(ActionCreator.decrementTime()),
  onSignIn: (email, password) => dispatch(UserOperation.signIn(email, password)),
  onReplayClick: () => dispatch(ActionCreator.reset()),
});

export {App};

export default connect(mapStateToProps, mapDispatchToProps)(App);
