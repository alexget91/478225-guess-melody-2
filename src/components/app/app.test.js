import React from "react";
import renderer from "react-test-renderer";
import {App} from "./app";
import createAudioMock from "../../common/test-stubs";
import {MemoryRouter} from "react-router-dom";
import Path from "../../common/path";

const mockGenreQuestion = {
  type: `genre`,
  genre: ``,
  answers: [{
    genre: ``,
    src: ``,
  }],
};
const mockArtistQuestion = {
  type: `artist`,
  song: {
    artist: ``,
    src: ``,
  },
  answers: [{
    artist: ``,
    picture: ``,
  }],
};

describe(`App correctly renders after relaunch`, () => {
  it(`Welcome screen`, () => {
    const tree = renderer
      .create(<MemoryRouter><App
        step={-1}
        maxMistakes={1}
        gameTime={1}
        questions={[]}
        onWelcomeScreenClick={jest.fn()}
        onSignIn={jest.fn()}
        onReplayClick={jest.fn()}
      /></MemoryRouter>)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Genre question screen`, () => {
    const tree = renderer
      .create(<MemoryRouter><App
        step={0}
        mistakes={0}
        gameTime={1}
        questions={[{
          type: `genre`,
          genre: ``,
          answers: [{
            genre: ``,
            src: ``,
          }],
        }]}
        onTimeChange={jest.fn()}
        onSignIn={jest.fn()}
        onReplayClick={jest.fn()}
      /></MemoryRouter>, createAudioMock())
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Artist question screen`, () => {
    const tree = renderer
      .create(<MemoryRouter><App
        step={0}
        mistakes={0}
        gameTime={1}
        questions={[mockArtistQuestion]}
        onTimeChange={jest.fn()}
        onSignIn={jest.fn()}
        onReplayClick={jest.fn()}
      /></MemoryRouter>, createAudioMock())
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Win screen`, () => {
    const tree = renderer
      .create(<MemoryRouter><App
        step={100}
        mistakes={0}
        gameTime={1}
        questions={[]}
        onTimeChange={jest.fn()}
        onSignIn={jest.fn()}
        onReplayClick={jest.fn()}
      /></MemoryRouter>)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Fail screen (time is over)`, () => {
    const tree = renderer
      .create(<MemoryRouter><App
        step={0}
        mistakes={0}
        gameTime={0}
        questions={[mockGenreQuestion]}
        onTimeChange={jest.fn()}
        onSignIn={jest.fn()}
        onReplayClick={jest.fn()}
      /></MemoryRouter>)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Fail screen (mistakes)`, () => {
    const tree = renderer
      .create(<MemoryRouter><App
        step={0}
        mistakes={10}
        maxMistakes={1}
        gameTime={1}
        questions={[mockGenreQuestion]}
        onTimeChange={jest.fn()}
        onSignIn={jest.fn()}
        onReplayClick={jest.fn()}
      /></MemoryRouter>)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Authorization screen`, () => {
    const tree = renderer
      .create(<MemoryRouter initialEntries={[Path.AUTH]}><App
        step={0}
        mistakes={0}
        gameTime={1}
        questions={[]}
        onTimeChange={jest.fn()}
        onSignIn={jest.fn()}
        onReplayClick={jest.fn()}
      /></MemoryRouter>)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
