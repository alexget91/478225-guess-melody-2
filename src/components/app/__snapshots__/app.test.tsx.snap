// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App correctly renders after relaunch Artist question screen 1`] = `
<section
  className="game game--artist"
>
  <header
    className="game__header"
  >
    <a
      className="game__back"
      href="#"
    >
      <span
        className="visually-hidden"
      >
        Сыграть ещё раз
      </span>
      <img
        alt="Угадай мелодию"
        className="game__logo"
        src="img/melody-logo-ginger.png"
      />
    </a>
    <svg
      className="timer"
      viewBox="0 0 780 780"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        className="timer__line"
        cx="390"
        cy="390"
        r="370"
        style={
          Object {
            "filter": "url(#blur); transform: rotate(-90deg) scaleY(-1); transform-origin: center",
          }
        }
      />
    </svg>
    <div
      className="timer__value"
    >
      <span
        className="timer__mins"
      >
        00
      </span>
      <span
        className="timer__dots"
      >
        :
      </span>
      <span
        className="timer__secs"
      >
        01
      </span>
    </div>
    <div
      className="game__mistakes"
    >
      
    </div>
  </header>
  <section
    className="game__screen"
  >
    <h2
      className="game__title"
    >
      Кто исполняет эту песню?
    </h2>
    <div
      className="game__track"
    >
      <div
        className="track"
      >
        <button
          className="track__button track__button--play js-track-button"
          disabled={true}
          onClick={[Function]}
          type="button"
        />
        <div
          className="track__status"
        >
          <audio />
        </div>
      </div>
    </div>
    <form
      className="game__artist"
    >
      <div
        className="artist"
      >
        <input
          className="artist__input visually-hidden js-answer-input"
          data-index={0}
          id="answer-0"
          name="answer"
          onChange={[Function]}
          type="radio"
          value="answer-0"
        />
        <label
          className="artist__name"
          htmlFor="answer-0"
        >
          <img
            alt=""
            className="artist__picture"
            src=""
          />
          
        </label>
      </div>
    </form>
  </section>
</section>
`;

exports[`App correctly renders after relaunch Authorization screen 1`] = `
<section
  className="login"
>
  <div
    className="login__logo"
  >
    <img
      alt="Угадай мелодию"
      height="83"
      src="img/melody-logo.png"
      width="186"
    />
  </div>
  <h2
    className="login__title"
  >
    Необходима авторизация
  </h2>
  <p
    className="login__text"
  >
    Представьтесь!
  </p>
  <form
    action=""
    className="login__form js-login-form"
    onSubmit={[Function]}
  >
    <p
      className="login__field"
    >
      <label
        className="login__label"
        htmlFor="name"
      >
        Логин
      </label>
      <input
        className="login__input js-email-input"
        id="name"
        name="name"
        required={true}
        type="email"
      />
    </p>
    <p
      className="login__field"
    >
      <label
        className="login__label"
        htmlFor="password"
      >
        Пароль
      </label>
      <input
        className="login__input js-password-input"
        id="password"
        name="password"
        required={true}
        type="password"
      />
      <span
        className="login__error"
      >
        Неверный пароль
      </span>
    </p>
    <button
      className="login__button button"
      type="submit"
    >
      Войти
    </button>
  </form>
</section>
`;

exports[`App correctly renders after relaunch Fail screen (mistakes) 1`] = `
<section
  className="result"
>
  <div
    className="result__logo"
  >
    <img
      alt="Угадай мелодию"
      height="83"
      src="img/melody-logo.png"
      width="186"
    />
  </div>
  <h2
    className="result__title"
  >
    Какая жалость!
  </h2>
  <p
    className="result__total result__total--fail"
  >
    У вас закончились все попытки. Ничего, повезёт в следующий раз!
  </p>
  <a
    className="replay"
    href="/"
    onClick={[Function]}
  >
    Попробовать ещё раз
  </a>
</section>
`;

exports[`App correctly renders after relaunch Fail screen (time is over) 1`] = `
<section
  className="result"
>
  <div
    className="result__logo"
  >
    <img
      alt="Угадай мелодию"
      height="83"
      src="img/melody-logo.png"
      width="186"
    />
  </div>
  <h2
    className="result__title"
  >
    Увы и ах!
  </h2>
  <p
    className="result__total result__total--fail"
  >
    Время вышло! Вы не успели отгадать все мелодии
  </p>
  <a
    className="replay"
    href="/"
    onClick={[Function]}
  >
    Попробовать ещё раз
  </a>
</section>
`;

exports[`App correctly renders after relaunch Genre question screen 1`] = `
<section
  className="game game--genre"
>
  <header
    className="game__header"
  >
    <a
      className="game__back"
      href="#"
    >
      <span
        className="visually-hidden"
      >
        Сыграть ещё раз
      </span>
      <img
        alt="Угадай мелодию"
        className="game__logo"
        src="img/melody-logo-ginger.png"
      />
    </a>
    <svg
      className="timer"
      viewBox="0 0 780 780"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        className="timer__line"
        cx="390"
        cy="390"
        r="370"
        style={
          Object {
            "filter": "url(#blur); transform: rotate(-90deg) scaleY(-1); transform-origin: center",
          }
        }
      />
    </svg>
    <div
      className="timer__value"
    >
      <span
        className="timer__mins"
      >
        00
      </span>
      <span
        className="timer__dots"
      >
        :
      </span>
      <span
        className="timer__secs"
      >
        01
      </span>
    </div>
    <div
      className="game__mistakes"
    >
      
    </div>
  </header>
  <section
    className="game__screen"
  >
    <h2
      className="game__title"
    >
      Выберите 
      
       треки
    </h2>
    <form
      className="game__tracks js-answer-form"
      onSubmit={[Function]}
    >
      <div
        className="track"
      >
        <button
          className="track__button track__button--play js-track-button"
          disabled={true}
          onClick={[Function]}
          type="button"
        />
        <div
          className="track__status"
        >
          <audio />
        </div>
        <div
          className="game__answer"
        >
          <input
            checked={false}
            className="game__input visually-hidden"
            data-index={0}
            id="answer-0"
            name="answer"
            onChange={[Function]}
            type="checkbox"
            value="answer-0"
          />
          <label
            className="game__check"
            htmlFor="answer-0"
          >
            Отметить
          </label>
        </div>
      </div>
      <button
        className="game__submit button"
        type="submit"
      >
        Ответить
      </button>
    </form>
  </section>
</section>
`;

exports[`App correctly renders after relaunch Welcome screen 1`] = `
<section
  className="welcome"
>
  <div
    className="welcome__logo"
  >
    <img
      alt="Угадай мелодию"
      height="83"
      src="img/melody-logo.png"
      width="186"
    />
  </div>
  <button
    className="welcome__button js-game-start"
    onClick={[MockFunction]}
  >
    <span
      className="visually-hidden"
    >
      Начать игру
    </span>
  </button>
  <h2
    className="welcome__rules-title"
  >
    Правила игры
  </h2>
  <p
    className="welcome__text"
  >
    Правила просты:
  </p>
  <ul
    className="welcome__rules-list"
  >
    <li>
      За 
      0.016666666666666666
       минут нужно ответить на все вопросы.
    </li>
    <li>
      Можно допустить 
      1
       ошибки.
    </li>
  </ul>
  <p
    className="welcome__text"
  >
    Удачи!
  </p>
</section>
`;

exports[`App correctly renders after relaunch Win screen 1`] = `
<section
  className="login"
>
  <div
    className="login__logo"
  >
    <img
      alt="Угадай мелодию"
      height="83"
      src="img/melody-logo.png"
      width="186"
    />
  </div>
  <h2
    className="login__title"
  >
    Вы настоящий меломан!
  </h2>
  <p
    className="login__total"
  >
    За 3 минуты и 25 секунд вы набрали 12 баллов (8 быстрых), совершив 3 ошибки
  </p>
  <p
    className="login__text"
  >
    Хотите сравнить свой результат с предыдущими попытками? Представьтесь!
  </p>
  <form
    action=""
    className="login__form js-login-form"
    onSubmit={[Function]}
  >
    <p
      className="login__field"
    >
      <label
        className="login__label"
        htmlFor="name"
      >
        Логин
      </label>
      <input
        className="login__input js-email-input"
        id="name"
        name="name"
        required={true}
        type="email"
      />
    </p>
    <p
      className="login__field"
    >
      <label
        className="login__label"
        htmlFor="password"
      >
        Пароль
      </label>
      <input
        className="login__input js-password-input"
        id="password"
        name="password"
        required={true}
        type="password"
      />
      <span
        className="login__error"
      >
        Неверный пароль
      </span>
    </p>
    <button
      className="login__button button"
      type="submit"
    >
      Войти
    </button>
  </form>
  <a
    className="replay"
    href="/"
    onClick={[Function]}
  >
    Сыграть ещё раз
  </a>
</section>
`;
